pact_testgen/broker.py

```diff
@@ -0,0 +1,93 @@
+import urllib
+from typing import Any, Dict, Optional, Tuple
+
+import requests
+from pydantic import BaseSettings, Field
+from pydantic.env_settings import InitSettingsSource, SettingsSourceCallable
+
+from .models import Pact
+
+
+# For broker settings, we want to allow passing None for username
+# and password, and revert to loading from the environment. However,
+# we also want to prioritize username and password if they are passed
+# as init kwargs, but are not None. This pattern simplifies CLI option
+# handling, at the expense of additional complexity here.
+
+
+class InitSettingsSourceIgnoreNone(InitSettingsSource):
+    """
+    An InitSettingsSource that ignores arguments
+    with a value of None.
+    """
+
+    def __init__(self, init_kwargs: Dict[str, Any]):
+        self.init_kwargs = {k: v for k, v in init_kwargs.items() if v is not None}
+        print(f"InitSettingsSourceIgnoreNone.init_kwargs={self.init_kwargs}")
```

Comment:
I got rid of this, but kept the `Config.customise_sources` override as a sort of happy medium.

Reply:
I think I'm going to explore refactoring the CLI code from argparse to Click, which supports retrieving options from env vars. Once done, these config `BaseSettings` classes can be replaced by standard `BaseModel`, and simply accept (or not) the arguments given to them.

We'll do this before adding support for token auth, because I think with Click the logic for determining which auth class to use will be simpler. If we added token auth support right now, we have weird possibilities like maybe you have env vars set for token auth, but pass a username + password as CLI options. We could probably just decide + document which would take precedence, but the logic ends up being split between the argparse code + the settings classes, and it's difficult to reason about, let alone test.